parser/prism/Translator.cc

```diff
@@ -0,0 +1,1059 @@
+        case PM_CALL_TARGET_NODE:
+        case PM_CAPTURE_PATTERN_NODE:
+        case PM_CASE_MATCH_NODE:
+        case PM_CLASS_VARIABLE_TARGET_NODE:
+        case PM_CONSTANT_AND_WRITE_NODE:
+        case PM_CONSTANT_OPERATOR_WRITE_NODE:
+        case PM_CONSTANT_OR_WRITE_NODE:
+        case PM_CONSTANT_PATH_AND_WRITE_NODE:
+        case PM_CONSTANT_PATH_OPERATOR_WRITE_NODE:
+        case PM_CONSTANT_PATH_OR_WRITE_NODE:
+        case PM_CONSTANT_PATH_TARGET_NODE:
+        case PM_CONSTANT_TARGET_NODE:
+        case PM_DEFINED_NODE:
+        case PM_EMBEDDED_VARIABLE_NODE:
+        case PM_ENSURE_NODE:
+        case PM_FIND_PATTERN_NODE:
+        case PM_FLIP_FLOP_NODE:
+        case PM_FOR_NODE:
+        case PM_GLOBAL_VARIABLE_TARGET_NODE:
+        case PM_HASH_PATTERN_NODE:
+        case PM_IMAGINARY_NODE:
+        case PM_IMPLICIT_NODE:
+        case PM_IMPLICIT_REST_NODE:
+        case PM_IN_NODE:
+        case PM_INDEX_TARGET_NODE:
+        case PM_INSTANCE_VARIABLE_TARGET_NODE:
+        case PM_INTERPOLATED_MATCH_LAST_LINE_NODE:
+        case PM_INTERPOLATED_REGULAR_EXPRESSION_NODE:
+        case PM_INTERPOLATED_SYMBOL_NODE:
+        case PM_INTERPOLATED_X_STRING_NODE:
+        case PM_IT_LOCAL_VARIABLE_READ_NODE: {
+            // See Prism::ParserStorage::ParsedRubyVersion
+            unreachable("The `it` keyword was introduced in Ruby 3.4, which isn't supported by Sorbet yet.");
+        }
+        case PM_IT_PARAMETERS_NODE: {
+            // See Prism::ParserStorage::ParsedRubyVersion
+            unreachable("The `it` keyword was introduced in Ruby 3.4, which isn't supported by Sorbet yet.");
+        }
+        case PM_LAMBDA_NODE:
+        case PM_MATCH_LAST_LINE_NODE:
+        case PM_MATCH_PREDICATE_NODE:
+        case PM_MATCH_REQUIRED_NODE:
+        case PM_MATCH_WRITE_NODE:
+        case PM_MISSING_NODE:
+        case PM_MULTI_TARGET_NODE:
+        case PM_NO_KEYWORDS_PARAMETER_NODE:
+        case PM_NUMBERED_PARAMETERS_NODE:
+        case PM_NUMBERED_REFERENCE_READ_NODE:
+        case PM_PINNED_EXPRESSION_NODE:
+        case PM_PINNED_VARIABLE_NODE:
+        case PM_POST_EXECUTION_NODE:
+        case PM_PRE_EXECUTION_NODE:
+        case PM_RESCUE_MODIFIER_NODE:
+        case PM_RESCUE_NODE:
+        case PM_SHAREABLE_CONSTANT_NODE:
+        case PM_UNDEF_NODE:
+        case PM_X_STRING_NODE:
+        case PM_SCOPE_NODE:
+            auto type_id = PM_NODE_TYPE(node);
+            auto type_name = pm_node_type_to_str(type_id);
+
+            fmt::memory_buffer buf;
+            fmt::format_to(std::back_inserter(buf), "Unimplemented node type {} (#{}).", type_name, type_id);
+            std::string s = fmt::to_string(buf);
+
+            auto fakeLocation = core::LocOffsets{0, 1};
+
+            return make_unique<parser::String>(fakeLocation, gs.enterNameUTF8(s));
+    }
+}
+
+std::unique_ptr<parser::Node> Translator::translate(const Node &node) {
+    return translate(node.get_raw_node_pointer());
+}
```

Comment:
You could consider moving this into the header, which gives it a better chance of inlining.